### PR TITLE
[Storage] Support Blob Service Properties in SRP

### PR DESCRIPTION
**Pull request by blueww**
## Description

This PR is to support manage Storage Blob Service Properties in SRP mode.
It target for 4/6 release.

The design already pass review in https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/263 

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/master/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (Microsoft internal partners only)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by blueww** [comment]
@markcowl , @maddieclayton , @cormacpayne 
could you share the staticAnalysis error in build?
Besides that, can we show staticAnalysis error in build log, or upload the output to some public blob that can share?
This can make the build failure investigation much more efficiently.

**Comment by blueww** [comment]
Add a static analysis exception for parameter name "RetentionDays" of cmdlet "Enable-AzStorageBlobDeleteRetentionPolicy".
This parameter name is aligned with the related parameter name of "Enable-AzStorageDeleteRetentionPolicy" https://github.com/Azure/azure-powershell/blob/3ef406acd59be744bcc345631b10f464a1d53bf9/src/Storage/Storage/Common/Cmdlet/EnableAzureStorageDeleteRetentionPolicy.cs#L32. The parameter name in this comlet has been reviewed and agreed with Mark before.

File: src/Storage/Storage.Management/Blob/DisableAzureStorageBlobDeleteRetentionPolicy.cs [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        [Alias(AccountNameAlias)]
+        [ValidateNotNullOrEmpty]
+        public string StorageAccountName { get; set; }
+
+        [Parameter(Mandatory = true,
+            HelpMessage = "Storage account object",
+            ValueFromPipeline = true,
+            ParameterSetName = AccountObjectParameterSet)]
+        [ValidateNotNullOrEmpty]
+        public PSStorageAccount StorageAccount { get; set; }
```
Comment (markcowl) [inline src/Storage/Storage.Management/Blob/DisableAzureStorageBlobDeleteRetentionPolicy.cs]: If you have this parameter set (that is a parameter set that takes PSStorageAccount), you should not have ValueFromPipelineByProeprtyName for the other properties.

File: src/Storage/Storage.Management/Blob/DisableAzureStorageBlobDeleteRetentionPolicy.cs [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+        [ValidateNotNullOrEmpty]
+        public string ResourceGroupName { get; set; }
+
+        [Parameter(
+            Position = 1,
+            Mandatory = true,
+            HelpMessage = "Storage Account Name.",
+           ParameterSetName = AccountNameParameterSet)]
+        [ResourceNameCompleter("Microsoft.Storage/storageAccounts", nameof(ResourceGroupName))]
+        [Alias(AccountNameAlias)]
```
Comment (markcowl) [inline src/Storage/Storage.Management/Blob/DisableAzureStorageBlobDeleteRetentionPolicy.cs]: Is this alias 'Name'?

File: src/Storage/Storage.Management/Blob/EnableAzureStorageBlobDeleteRetentionPolicy.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+        [Alias(AccountNameAlias)]
+        [ValidateNotNullOrEmpty]
+        public string StorageAccountName { get; set; }
+
+        [Parameter(Mandatory = true,
+            HelpMessage = "Storage account object",
+            ValueFromPipeline = true,
+            ParameterSetName = AccountObjectParameterSet)]
+        [ValidateNotNullOrEmpty]
+        public PSStorageAccount StorageAccount { get; set; }
```
Comment (markcowl) [inline src/Storage/Storage.Management/Blob/EnableAzureStorageBlobDeleteRetentionPolicy.cs]: same comment - this parameter should take values from the pipeline, and ther should be no parameters that take values from pieline by property name.

File: src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        [Alias(AccountNameAlias)]
+        [ValidateNotNullOrEmpty]
+        public string StorageAccountName { get; set; }
+
+        [Parameter(Mandatory = true,
+            HelpMessage = "Storage account object",
+            ValueFromPipeline = true,
+            ParameterSetName = AccountObjectParameterSet)]
+        [ValidateNotNullOrEmpty]
+        public PSStorageAccount StorageAccount { get; set; }
```
Comment (markcowl) [inline src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs]: Same comment - there shoudl be no parameters with 'ValueFromPiepelineByProeprtyName'

File: src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            Position = 0,
+            Mandatory = true,
+            ValueFromPipelineByPropertyName = true,
+            HelpMessage = "Blob Service Properties Resource Id.",
+           ParameterSetName = PropertiesResourceIdParameterSet)]
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Display ServiceProperties")]
+        public SwitchParameter PassThru { get; set; }
```
Comment (markcowl) [inline src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs]: Why does this have a passthru parameter?  The Get should always return the BlobServiceProperties.

File: src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters;
+    using Microsoft.Azure.Management.Internal.Resources.Utilities.Models;
+
+    /// <summary>
+    /// Modify Azure Storage service properties
+    /// </summary>
+    [Cmdlet("Get", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + StorageBlobServiceProperty, DefaultParameterSetName = AccountNameParameterSet), OutputType(typeof(PSBlobServiceProperties))]
```
Comment (markcowl) [inline src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs]: Shouldn't thsi just be 'Get-AzStorageBlobService'?

File: src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Management.Automation;
+    using Microsoft.Azure.Commands.ResourceManager.Common.ArgumentCompleters;
+    using Microsoft.Azure.Management.Internal.Resources.Utilities.Models;
+
+    /// <summary>
+    /// Modify Azure Storage service properties
+    /// </summary>
+    [Cmdlet("Update", ResourceManager.Common.AzureRMConstants.AzureRMPrefix + StorageBlobServiceProperty, SupportsShouldProcess = true, DefaultParameterSetName = AccountNameParameterSet), OutputType(typeof(PSBlobServiceProperties))]
```
Comment (markcowl) [inline src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs]: Same comment, shouldn't this be Update-AzStorageBlobService?

File: src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+        [Alias(AccountNameAlias)]
+        [ValidateNotNullOrEmpty]
+        public string StorageAccountName { get; set; }
+
+        [Parameter(Mandatory = true,
+            HelpMessage = "Storage account object",
+            ValueFromPipeline = true,
+            ParameterSetName = AccountObjectParameterSet)]
+        [ValidateNotNullOrEmpty]
+        public PSStorageAccount StorageAccount { get; set; }
```
Comment (markcowl) [inline src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs]: Same comment - this is the only parameter that should take its value from the pipeline

File: src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+        [ValidateNotNullOrEmpty]
+        public string ResourceId { get; set; }
+
+        [Parameter(Mandatory = false, HelpMessage = "Default Service Version to Set")]
+        [ValidateNotNull]
+        public string DefaultServiceVersion { get; set; }
+
+        [Parameter(Mandatory = true, HelpMessage = "Blob service properties object to Set", ParameterSetName = AccountNamePropertiesObjectParameterSet, ValueFromPipeline = true)]
+        [ValidateNotNullOrEmpty]
+        public PSBlobServiceProperties BlobServiceProperty { get; set; }
```
Comment (markcowl) [inline src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs]: Shouldn't this be the parameter that takes its value from the pipeline?  In that case, the verb for this cmdlet should eb 'Set'.  'Update' only makes sense if you are using PATCH semantcis, meaning that you can set one blob service property and leave others unchanged.

File: src/Storage/Storage.Management/Blob/DisableAzureStorageBlobDeleteRetentionPolicy.cs (hunk 2 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/DisableAzureStorageBlobDeleteRetentionPolicy.cs]: the alias is "AccountName"
`protected const string AccountNameAlias = "AccountName";`

File: src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs (hunk 5 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs]: Removed. Thanks for catching it!

File: src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs (hunk 6 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs]: Blob Service support Container, blob, blob service priperties. Add properties might be more specific, and this is also aligned with dataplane cmdlet name: Get-AzureStorageServiceProperties

File: src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs (hunk 7 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs]: Blob Service support Container, blob, blob service priperties. Add "properties" might be more specific, to let user know the blob/container won't updated. only service properties are updated.
And this is also aligned with dataplane cmdlet name: Update-AzureStorageServiceProperties

File: src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs (hunk 9 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs]: We can set one blob service property and leave others unchanged, so "Update" might be more proper. 
Actually, this is also aligned with dataplane cmdlet name: Update-AzureStorageServiceProperties. And "Update" verb is suggest by you when design the dataplane cmdlet

Comment (blueww) [reply src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs]: As we talked, I have removed parameter "PSBlobServiceProperties BlobServiceProperty" from this cmdlet. We might can add a Set-* cmdlet for this in the future, but not in this release.

File: src/Storage/Storage.Management/Blob/DisableAzureStorageBlobDeleteRetentionPolicy.cs (hunk 1 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/DisableAzureStorageBlobDeleteRetentionPolicy.cs]: As we discussed, the only parameter has ValueFromPipelineByProeprtyName os "-ResourceId" and this is acceptable. 
As we discussed, I have change the parameter name in all 4 new cmdlets to "ResourceId", and all will accept both Storage account resourceId and Blob Service ResourceId.

File: src/Storage/Storage.Management/Blob/EnableAzureStorageBlobDeleteRetentionPolicy.cs (hunk 3 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/EnableAzureStorageBlobDeleteRetentionPolicy.cs]: As we discussed, the only parameter has ValueFromPipelineByProeprtyName os "-ResourceId" and this is acceptable. 
As we discussed, I have change the parameter name in all 4 new cmdlets to "ResourceId", and all will accept both Storage account resourceId and Blob Service ResourceId.

File: src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs (hunk 4 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs]: As we discussed, the only parameter has ValueFromPipelineByProeprtyName os "-ResourceId" and this is acceptable. 
As we discussed, I have change the parameter name in all 4 new cmdlets to "ResourceId", and all will accept both Storage account resourceId and Blob Service ResourceId.

File: src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs (hunk 6 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/GetAzureStorageBlobServiceProperties.cs]: As we discussed ,will still use Get-AzureStorageServiceProperty

File: src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs (hunk 7 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs]: As we discussed ,will still use Update-AzureStorageServiceProperty

File: src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs (hunk 8 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/UpdateAzureStorageBlobServiceProperties.cs]: As we discussed, the only parameter has ValueFromPipelineByProeprtyName os "-ResourceId" and this is acceptable.
As we discussed, I have change the parameter name in all 4 new cmdlets to "ResourceId", and all will accept both Storage account resourceId and Blob Service ResourceId.

**Comment by blueww** [comment]
@markcowl 
I have updated the PR per our discusstion:
1. Make the parameter name to -ResouceId, and support both Storage account Id and Blob Service properties Id
2. Remove "PSBlobServiceProperties BlobServiceProperty" parameter from Update-AzureStorageBlobServiceProperty. so we can keep the name as Update-*
3. remove -passthru  from Get-* cmdlet
4. other changelog and help related change.

Please help to check and merge it if no other issues.
Thanks!

**Comment by markcowl** [comment]
@blueww One remaining issue, otherwise lgtm

File: src/Storage/Storage.Management/Blob/DisableAzureStorageBlobDeleteRetentionPolicy.cs (hunk 2 above)
Comment (blueww) [reply src/Storage/Storage.Management/Blob/DisableAzureStorageBlobDeleteRetentionPolicy.cs]: Fixed by add alias "Name".
See commit https://github.com/Azure/azure-powershell/pull/8874/commits/8749ddd484c4ae299796c4019ee0843c90249192

**Comment by blueww** [comment]
@markcowl 
I have fixed the last comment from you, in commit https://github.com/Azure/azure-powershell/pull/8874/commits/8749ddd484c4ae299796c4019ee0843c90249192
Please help to check and merge the PR.

**Comment by blueww** [comment]
@markcowl 
It seems the build failed, but not caused by my code change:
> The command "sudo apt-get update" failed and exited with 100 during .
Would you please help to fix the build? 
Is there any easy way to rerun build of the PR?

**Pull request merged**